File: Java/90. Subsets II.java
class Solution {
    private List<List<Integer>> ans = new ArrayList<>();
    
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        backtrack(nums, 0, new LinkedList<Integer>());
        return ans;
    }
    
    private void backtrack (int[]nums, int start, LinkedList<Integer> cur) {
        ans.add(new ArrayList<>(cur));
        for (int i=start; i<nums.length; i++) {
            /* found a duplicate */
            if (i != start && nums[i] == nums[i-1]) {
                continue;
            }
            
            cur.add(nums[i]);
            backtrack(nums, i+1, cur);
            cur.removeLast();
        }
    }
}

File: Python3/90. Subsets II.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        ans = []
        nums.sort()
        
        def backtrack(start=0, cur=[]):
            ans.append(cur[:])
            if start == len(nums):
                return
            
            for i in range(start, len(nums)):
                # found a duplicate
                if i > start and nums[i] == nums[i-1]:
                    continue
                    
                cur.append(nums[i])
                #print(nums[i], cur)
                backtrack(i+1, cur)
                cur.pop()
                
        backtrack()
        return ans